fetcher/bundle.go

```diff
@@ -0,0 +1,206 @@
+	logrus.Infof("Sending bundle creation request to %s", b.Node.IP)
+
+	url := fmt.Sprintf("http://%s:%d/system/health/v1/diagnostics/%s", b.Node.IP, b.port, b.BundleID)
+	fullURL, err := util.UseTLSScheme(url, b.forceTLS)
+
+	if err != nil {
+		return err
+	}
+
+	type payload struct {
+		Type string `json:"type"`
+	}
+
+	body, err := json.Marshal(payload{
+		Type: "LOCAL",
+	})
+	if err != nil {
+		return err
+	}
+
+	request, err := http.NewRequest("PUT", fullURL, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	request.WithContext(ctx)
+
+	resp, err := b.Client.Do(request)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected HTTP status code received: %d", resp.StatusCode)
+	}
+
+	// start update checker
+	go statusUpdateCheck(ctx, b, requestFinished)
+
+	return nil
+}
+
+func (b *BundleRequest) checkStatus(ctx context.Context) RequestStatus {
+	status := RequestStatus{
+		Request: b,
+	}
+
+	logrus.Infof("Checking status of bundle creation on node %s", b.Node.IP)
+	url := fmt.Sprintf("http://%s:%d/system/health/v1/diagnostics/%s", b.Node.IP, b.port, b.BundleID)
+
+	fullURL, err := util.UseTLSScheme(url, b.forceTLS)
+	if err != nil {
+		// TODO
+	}
+
+	request, err := http.NewRequest("GET", fullURL, nil)
+	if err != nil {
+		status.Err = err
+		return status
+	}
+
+	request.WithContext(ctx)
+
+	resp, err := b.Client.Do(request)
+	if err != nil {
+		status.Err = err
+		return status
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == 200 {
+		var body rest.Bundle
+
+		err = json.NewDecoder(resp.Body).Decode(&body)
+		if err != nil {
+			status.Err = fmt.Errorf("Error decoding status response: %s", err)
+		}
+
+		if body.Status == rest.Done || body.Status == rest.Canceled {
+			status.Finished = true
+		} else {
+			status.Finished = false
+		}
+	} else if resp.StatusCode == 404 {
+		status.Err = fmt.Errorf("Bundle ID %s not known to agent", b.BundleID)
+		status.Finished = true
+	} else {
+		status.Err = fmt.Errorf("received unexpected status code %d", resp.StatusCode)
+		status.Finished = true
+	}
+
+	return status
+}
+
+// Download will download the bundle file created by this Bundle Request,
+// successful download will result in the data being put into b.Data
+func (b *BundleRequest) Download(ctx context.Context) error {
+	logrus.Infof("downloading bundle %s from node %s", b.BundleID, b.Node.IP)
+	url := fmt.Sprintf("http://%s:%d/system/health/v1/diagnostics/%s/file", b.Node.IP, b.port, b.BundleID)
+
+	fullURL, err := util.UseTLSScheme(url, b.forceTLS)
+	if err != nil {
+		return err
+	}
+
+	request, err := http.NewRequest("GET", fullURL, nil)
+	if err != nil {
+		return err
+	}
+
+	request.WithContext(ctx)
+
+	resp, err := b.Client.Do(request)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		body, err := ioutil.ReadAll(resp.Body)
```

Comment:
Storing bundle in memory is not the best idea. Bundle from local machine could be pretty big. I think we should write it on disk and then merge into a single bundle like we did in fetcher.go